### DISCUSSION TITLE
Typings for the JS sdk

### DISCUSSION THREAD

**Discussion by david-plugge** [pocketbase/pocketbase#2750] [Show and tell]
Hi everyone!

I´ve been using pocketbase for a few weeks now and i´m more than happy with literally everything about it.
One thing that bothered me though is that the typescript support is kinda lacking (compared to prisma/drizzle/kyzely/supabase).
Passing the collection types on every call is very repetitive and can be avoided.

So i tried to solve this issue and here is what i came up with: https://github.com/david-plugge/typed-pocketbase

I´d absolutely love some feedback!

Happy coding :)

**Comment by charbs-io** [comment]
I found this in the awesome collection I think. All I can say is thank you, it had made life a lot better 😀

**Comment by silenceofthewans** [comment]
I've been wanting this in JS SDK and this is a really good start thank you! Some things I noticed: 

- There is no intellisense description about operators and don't know if it is possible using strings union because I could only make it work with an enum when I was using a tuple as types helper. 

- ~~Previosuly, I could also set a type of a response from API which is initially unknown to a CollectionResponse type that pocketbase-typegen provides but using your package this stops working and the type reads as TypedRecord<....> so I tried using TypedRecord<CollectionResponse> but that doesn't work because TypedRecord isn't exported.~~ I'm also a typescript beginner so don't if this is how it's supposed to work but currently I see a lot of type clutter using this package like one of my collection svelte store array reads as instead of just ProjectRecord[]:
 ```
 TypedRecord<ProjectRecord & SystemFields, "name" | "id" | "update" | "owner", {
    name?: string | undefined;
    id: string;
    update: string;
    owner: string;
 }>[]
 ```
So, would love if this get a little bit more simplified.

- The "update" field should be changed to "updated" ~~and I wonder if this also the cause of my response type error since the BaseSystemFields in pocketbase-typegen use "updated" instead of typed-pocketbase "update".~~

Eventually, I would love to see something like this in the JS SDK itself instead of installing two different packages to get type safety but this is the next best option. 👍

EDIT: After some more scenarios and looking at source, their is no distinction between Record and Response as pocketbase-typegen creates. It also doesn't use `BaseSystemFields` from pocketbase-typegen and creates it's own `SystemFields` interface which is missing expand and expand generic argument support that pocketbase-typegen provides like `ItemResponse<MyExpandedRecord>`. So, currently I'm going back to just passing generics but will give this a try again after the fixes.

**Reply by david-plugge** [reply]
Thanks for your feedback!

A lot of stuff is still missing, i just wanted to get this out to see if people are willing to use it.
I´m planning to add more stuff, like typesafe expand and the ability to overide the types.
The type readability should also improve with the next version.
Feel free to add more stuff to the list :)

**Reply by silenceofthewans** [reply]
That's Awesome! I'll be waiting for it 😊 and I think so will a lot of other people. Some stuff that I would love to see in no particular order:

- Using mostly the collection types that `pocketbase-typegen` provides as those can be auto-generated using something like [this ](https://github.com/pocketbase/pocketbase/discussions/1838) on collection changes. I guess this might not be possible in some cases and ability to override the types will be fine then.
- Using Record type (fields) when using filters or creating,updating,deleting records and properly setting the Response type (fields+BaseSystemFields) for all methods that return `Record | Record[]`
- Eventually when type safe `expand()` will be added it should only provide intellisense on fields that are actually expandable. pocketbase-typegen gives those a RecordIdString, RecordIdString[] type. These field's Response (or Responses in case of multi or nested expand) should then properly be typed into the Texpand generic of that collection's Response<Texpand = unknown> so expand can be accessed if it was passed in.

I think above will basically provide most of the desired type safety. 

P.S This package and pocketbase-typegen benefit from each other and could probably become one project so community only needs to install one to get type safety ✨ but that's up to you and @patmood
